Mathlib/CategoryTheory/Sigma/Basic.lean

```diff
@@ -0,0 +1,286 @@
+
+/-- If `q` when restricted to each subcategory `C i` agrees with `F i`, then `q` is isomorphic to
+`desc F`.
+-/
+def descUniq (q : (Σi, C i) ⥤ D) (h : ∀ i, incl i ⋙ q ≅ F i) : q ≅ desc F :=
+  (NatIso.ofComponents fun ⟨i, X⟩ => (h i).app X) <| by
+    rintro ⟨i, X⟩ ⟨_, _⟩ ⟨f⟩
+    apply (h i).hom.naturality f
+#align category_theory.sigma.desc_uniq CategoryTheory.Sigma.descUniq
+
+@[simp]
+theorem descUniq_hom_app (q : (Σi, C i) ⥤ D) (h : ∀ i, incl i ⋙ q ≅ F i) (i : I) (X : C i) :
+    (descUniq F q h).hom.app ⟨i, X⟩ = (h i).hom.app X :=
+  rfl
+#align category_theory.sigma.desc_uniq_hom_app CategoryTheory.Sigma.descUniq_hom_app
+
+@[simp]
+theorem descUniq_inv_app (q : (Σi, C i) ⥤ D) (h : ∀ i, incl i ⋙ q ≅ F i) (i : I) (X : C i) :
+    (descUniq F q h).inv.app ⟨i, X⟩ = (h i).inv.app X :=
+  rfl
+#align category_theory.sigma.desc_uniq_inv_app CategoryTheory.Sigma.descUniq_inv_app
+
+/--
+If `q₁` and `q₂` when restricted to each subcategory `C i` agree, then `q₁` and `q₂` are isomorphic.
+-/
+@[simps]
+def natIso {q₁ q₂ : (Σi, C i) ⥤ D} (h : ∀ i, incl i ⋙ q₁ ≅ incl i ⋙ q₂) : q₁ ≅ q₂ where
+  hom := natTrans fun i => (h i).hom
+  inv := natTrans fun i => (h i).inv
+#align category_theory.sigma.nat_iso CategoryTheory.Sigma.natIso
+
+end
+
+section
+
+variable (C) {J : Type w₂} (g : J → I)
+
+/-- A function `J → I` induces a functor `Σ j, C (g j) ⥤ Σ i, C i`. -/
+def map : (Σj : J, C (g j)) ⥤ Σi : I, C i :=
+  desc fun j => incl (g j)
+#align category_theory.sigma.map CategoryTheory.Sigma.map
+
+@[simp]
+theorem map_obj (j : J) (X : C (g j)) : (Sigma.map C g).obj ⟨j, X⟩ = ⟨g j, X⟩ :=
+  rfl
+#align category_theory.sigma.map_obj CategoryTheory.Sigma.map_obj
+
+@[simp]
+theorem map_map {j : J} {X Y : C (g j)} (f : X ⟶ Y) :
+    (Sigma.map C g).map (SigmaHom.mk f) = SigmaHom.mk f :=
+  rfl
+#align category_theory.sigma.map_map CategoryTheory.Sigma.map_map
+
+/-- The functor `Sigma.map C g` restricted to the subcategory `C j` acts as the inclusion of `g j`.
+-/
+@[simps]
+def inclCompMap (j : J) : incl j ⋙ map C g ≅ incl (g j) :=
+  Iso.refl _
+#align category_theory.sigma.incl_comp_map CategoryTheory.Sigma.inclCompMap
+
+variable (I)
+
+/-- The functor `Sigma.map` applied to the identity function is just the identity functor. -/
+@[simps]
+def mapId : map C (id : I → I) ≅ 𝟭 (Σi, C i) :=
+  natIso fun i => NatIso.ofComponents (fun X => Iso.refl _) (by aesop)
+#align category_theory.sigma.map_id CategoryTheory.Sigma.mapId
+
+-- Porting note: TODO: don't know how to fix the type errors here
+-- variable {I} {K : Type w₃}
+
+-- /-- The functor `sigma.map` applied to a composition is a composition of functors. -/
+-- @[simps]
+-- def mapComp (f : K → J) (g : J → I) : map (C ∘ g) f ⋙ (map C g : _) ≅ map C (g ∘ f) :=
+--   (descUniq _ _) fun k =>
+--     (isoWhiskerRight (inclCompMap (C ∘ g) f k) (map C g : _) : _) ≪≫ inclCompMap _ _ _
+
+-- #align category_theory.sigma.map_comp CategoryTheory.Sigma.mapComp
```

Comment:
This one will need fixing, though I'm not quite sure how either.

Reply:
I think I found the issue.